src/custom_token.rs

```diff
@@ -0,0 +1,534 @@
+// Not public API, but #[macro_export] _is_ documented
+#![doc(hidden)]
+
+#[cfg(feature = "parsing")]
+use token;
+#[cfg(feature = "parsing")]
+use buffer::Cursor;
+
+#[cfg(feature = "parsing")]
+pub trait CustomToken {
+    fn peek(cursor: Cursor) -> bool;
+    fn display() -> &'static str;
+}
+
+#[cfg(feature = "parsing")]
+impl<T: CustomToken> token::private::Sealed for T {}
+#[cfg(feature = "parsing")]
+impl<T: CustomToken> token::Token for T {
+    fn peek(cursor: Cursor) -> bool {
+        <Self as CustomToken>::peek(cursor)
+    }
+
+    fn display() -> &'static str {
+        <Self as CustomToken>::display()
+    }
+}
+
+// Not public API.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! punct_len {
+    ()       => ( 0usize );
+    (+)      => ( 1usize );
+    (+=)     => ( 2usize );
+    (&)      => ( 1usize );
+    (&&)     => ( 2usize );
+    (&=)     => ( 2usize );
+    (@)      => ( 1usize );
+    (!)      => ( 1usize );
+    (^)      => ( 1usize );
+    (^=)     => ( 2usize );
+    (:)      => ( 1usize );
+    (::)     => ( 2usize );
+    (,)      => ( 1usize );
+    (/)      => ( 1usize );
+    (/=)     => ( 2usize );
+    (.)      => ( 1usize );
+    (..)     => ( 2usize );
+    (...)    => ( 3usize );
+    (..=)    => ( 3usize );
+    (=)      => ( 1usize );
+    (==)     => ( 2usize );
+    (>=)     => ( 2usize );
+    (>)      => ( 1usize );
+    (<=)     => ( 2usize );
+    (<)      => ( 1usize );
+    (*=)     => ( 2usize );
+    (!=)     => ( 2usize );
+    (|)      => ( 1usize );
+    (|=)     => ( 2usize );
+    (||)     => ( 2usize );
+    (#)      => ( 1usize );
+    (?)      => ( 1usize );
+    (->)     => ( 2usize );
+    (<-)     => ( 2usize );
+    (%)      => ( 1usize );
+    (%=)     => ( 2usize );
+    (=>)     => ( 2usize );
+    (;)      => ( 1usize );
+    (<<)     => ( 2usize );
+    (<<=)    => ( 3usize );
+    (>>)     => ( 2usize );
+    (>>=)    => ( 3usize );
+    (*)      => ( 1usize );
+    (-)      => ( 1usize );
+    (-=)     => ( 2usize );
+    (~)      => ( 1usize );
+    (_)      => ( 1usize );
```

Comment:
Let's disallow underscore in custom punctuation. The compiler has transitioned to treating `_` as an identifier rather than punctuation because it is most commonly used in places where you would otherwise see an identifier, and it does not combine with other punctuation.